### ISSUE TITLE
feat(cdk/layout): Container breakpoint observer

### ISSUE THREAD

**Issue by SteveVanOpstal** [angular/components#26537]
### Feature Description

A breakpoint observer that observes the current component or component parent. Breakpoints on the component level allow you to create truly dynamic components.

For some time I've been using a library (that I created) called [angular-container-media-query](https://www.npmjs.com/package/angular-container-media-query) that does this in a suboptimal way. I was looking at ways to rework this and came to the conclusion that it would fit well within the `cdk/layout`.

The api for the `angular-container-media-query` library requires you to "register" a resize service:
```
import {MediaQuery, ObserveResizeService} from 'angular-container-media-query';

@Component({})
export class Component {
  @MediaQuery('(max-width: 400px)') @HostBinding('class.small') small = false;

  constructor(
    resize: ObserveResizeService,
    elementRef: ElementRef,
    changeDetector: ChangeDetectorRef
  ) {
    resize.register(this, elementRef, changeDetector);
  }
```

The api could more closely resemble the `BreakpointObserver`, but I'm not sure if that's possible:
```
import {ComponentBreakpointObserver} from '@angular/cdk/layout';

@Component({})
export class Component implements OnInit, OnDestroy {
  @HostBinding('class.small') small = false;

  small$ = this._componentBreakpointObserver.observe('(max-width: 400px)').pipe(
    map(({matches}) => matches)
  );

  constructor(
    private _componentBreakpointObserver: ComponentBreakpointObserver,
  ) {}

  ngOnInit() {
    this.small$.pipe(takeUntil(this.destroyed)).subscribe((small) => this.small = small);
  }

  ngOnDestroy() {
    this.destroyed.next();
    this.destroyed.complete();
  }
```

### Use Case

When you use for example a sidenav the content area resizes based on the sidenav being open or closed. A component within that content will better respond to size changes based on the content area size (or its own size) instead of listening to the breakpoints on the viewport.

I reused the dynamic sidenav example to show how the ContainerBreakpointObserver could work (and how it currently works with the `angular-container-media-query` library):
https://stackblitz.com/edit/angular-59mxnc

![image](https://user-images.githubusercontent.com/7734728/216007576-83e7492b-a39d-488f-a34a-6a8a3707a886.png)![image](https://user-images.githubusercontent.com/7734728/216007633-b87dbaa5-8737-4ee5-aeb4-861d3d13dd43.png)

**Comment by digeomel** [comment]
Wow, I came here looking to see what the Angular team is planning to do about container queries now that they're officially out and I came across this 😃 
I guess it goes without saying that we should gradually dismiss breakpoint observers on the whole viewport and start adding them on components.
I, for one, support any suggestion towards this end! 👍🏼 
Oh, and for now I will try your library @SteveVanOpstal , thanks!
